src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -204,151 +234,268 @@ private static VersionInfo GetVersionInfo()
             {
-                var getters = new ValueGetter<VBuffer<Float>>[Infos.Length];
-                for (int i = 0; i < Infos.Length; i++)
-                    getters[i] = cursor.GetGetter<VBuffer<Float>>(Infos[i].Source);
-                var val = default(VBuffer<Float>);
+                var getters = new ValueGetter<VBuffer<float>>[columns.Length];
+                for (int i = 0; i < columns.Length; i++)
+                    getters[i] = cursor.GetGetter<VBuffer<float>>(cols[i]);
+                var val = default(VBuffer<float>);
                 int irow = 0;
                 while (irow < maxActualRowCount && cursor.MoveNext())
                 {
-                    for (int i = 0; i < Infos.Length; i++)
+                    for (int i = 0; i < columns.Length; i++)
                     {
                         if (irow >= actualRowCounts[i] || columnData[i].Length == 0)
                             continue;
 
                         getters[i](ref val);
                         val.CopyTo(columnData[i], idxDst[i]);
-                        idxDst[i] += Infos[i].TypeSrc.ValueCount;
+                        idxDst[i] += srcTypes[i].ValueCount;
                     }
                     irow++;
                 }
 #if DEBUG
-                for (int i = 0; i < Infos.Length; i++)
+                for (int i = 0; i < columns.Length; i++)
                     ch.Assert(idxDst[i] == columnData[i].Length);
 #endif
             }
-
             return columnData;
         }
 
-        private void TrainModels(Float[][] columnData, int[] rowCounts, IChannel ch)
+        // Performs whitening training for each column separately. Notice that for both PCA and ZCA, _models and _invModels
+        // will have dimension input_vec_size x input_vec_size. In the getter, the matrix will be truncated to only keep
+        // PcaNum columns, and thus produce the desired output size.
+        private static void TrainModels(IHostEnvironment env, IChannel ch, float[][] columnData, int[] rowCounts,
+            ref float[][] models, ref float[][] invModels, ColumnType[] srcTypes, params ColumnInfo[] columns)
         {
-            Host.AssertValue(ch);
             ch.Assert(columnData.Length == rowCounts.Length);
 
-            for (int iinfo = 0; iinfo < Infos.Length; iinfo++)
+            for (int iinfo = 0; iinfo < columns.Length; iinfo++)
             {
-                var ex = _exes[iinfo];
+                var ex = columns[iinfo];
                 var data = columnData[iinfo];
                 int crow = rowCounts[iinfo];
-                int ccol = Infos[iinfo].TypeSrc.ValueCount;
+                int ccol = srcTypes[iinfo].ValueCount;
+
+                // If there is no training data, simply initialize the model matrices.
+                if (crow == 0)
+                {
+                    var matrixSize = ccol * ccol;
+                    models[iinfo] = new float[matrixSize];
```

Comment:
If there is no training data, shouldn't the model be an identity matrix?

Reply:
Yes I agree, I am turning them into identity matrices.

---
In reply to: [232023973](https://github.com/dotnet/machinelearning/pull/1452#discussion_r232023973) [](ancestors = 232023973)